sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+                ble_att_set_preferred_mtu(MTU_MAX_NIMBLE);
+                ble_gattc_exchange_mtu(_conn_handle, NULL, NULL);
+            }
+            else {
+                _status = STDIO_NIMBLE_DISCONNECTED;
+            }
+            break;
+
+        case BLE_GAP_EVENT_DISCONNECT:
+            _debug_printf("BLE_GAP_EVENT_DISCONNECT\n");
+            _status = STDIO_NIMBLE_DISCONNECTED;
+            break;
+
+        case BLE_GAP_EVENT_SUBSCRIBE:
+            _debug_printf("BLE_GAP_EVENT_SUBSCRIBE\n");
+            if (event->subscribe.attr_handle == _val_handle_stdout) {
+                if (event->subscribe.cur_indicate == 1) {
+                    _status = STDIO_NIMBLE_SUBSCRIBED;
+                }
+                else {
+                    _status = STDIO_NIMBLE_CONNECTED;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_NOTIFY_TX:
+            _debug_printf("BLE_GAP_EVENT_NOTIFY_TX\n");
+            if (event->notify_tx.indication == 1) {
+                if (event->notify_tx.status == BLE_HS_EDONE) {
+                    _status = STDIO_NIMBLE_SUBSCRIBED;
+                }
+                else if (event->notify_tx.status != 0) {
+                    _status = STDIO_NIMBLE_SUBSCRIBED;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_MTU:
+            _debug_printf("BLE_GAP_EVENT_MTU\n");
+            _current_max_mtu_payload = event->mtu.value - PDU_HEADERS_SIZE;
+            break;
+    }
+
+    return 0;
+}
+
+static int gatt_svr_chr_access_stdin(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg)
+{
+    (void) conn_handle;
+    (void) attr_handle;
+    (void) arg;
+
+    uint16_t om_len = OS_MBUF_PKTLEN(ctxt->om);
+
+    /* read sent data */
+    int rc = ble_hs_mbuf_to_flat(ctxt->om, _stdin_read_buf,
+            sizeof(_stdin_read_buf), &om_len);
+
+    isrpipe_write(&_isrpipe_stdin, _stdin_read_buf, om_len);
+
+    return rc;
+}
+
+void stdio_init(void)
+{
+#if IS_USED(MODULE_VFS)
+    vfs_bind_stdio();
+#endif
+
+#if IS_USED(MODULE_STDIO_NIMBLE_DEBUG)
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, NULL, NULL);
+#endif
+
+    ble_npl_callout_init(&_send_stdout_callout, nimble_port_get_dflt_eventq(),
+                         _send_stdout, NULL);
+}
+
+ssize_t stdio_read(void *buffer, size_t count)
+{
+    /* blocks until at least one character was read */
+    ssize_t res = isrpipe_read(&_isrpipe_stdin, buffer, count);
+
+#if IS_USED(MODULE_STDIO_NIMBLE_DEBUG)
+    mutex_lock(&_debug_print_mutex);
+    uart_write(STDIO_UART_DEV, (const uint8_t *)PREFIX_STDIN, strlen(PREFIX_STDIN));
+    uart_write(STDIO_UART_DEV, (const uint8_t *)buffer, res);
+    mutex_unlock(&_debug_print_mutex);
+#endif
+
+    return res;
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+#if IS_USED(MODULE_STDIO_NIMBLE_DEBUG)
+    mutex_lock(&_debug_print_mutex);
```

Comment:
Why are you using a mutex here btw?

Reply:
Without it messages could clash. E.g. I could get `OUT: s` instead of `STDOUT: s`. The `STD` gets simply lost. This does not happen when protecting `uart_write` by a mutex. I even had something like the following:
```
2021-12-07 23:29:41,382 # SBLE_GAP_EVENT_NOTIFY_TX
2021-12-07 23:29:41,383 # 2 bytes sent successfully
2021-12-07 23:29:41,383 # TDOUT: j
```
As you can see the original message `STDOUT: j` is broken into two parts.